app/src/main/java/io/homeassistant/companion/android/widgets/WidgetDynamicFieldAdapter.kt

```diff
@@ -0,0 +1,100 @@
+    }
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
+        val inflater = LayoutInflater.from(parent.context)
+
+        val dynamicFieldLayout = inflater.inflate(
+            io.homeassistant.companion.android.R.layout.widget_button_configure_dynamic_field,
+            parent,
+            false
+        )
+
+        return ViewHolder(dynamicFieldLayout)
+    }
+
+    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
+        val dynamicFieldLayout = holder.itemView
+        val autoCompleteTextView = dynamicFieldLayout.dynamic_autocomplete_textview
+        val context = dynamicFieldLayout.context
+
+        val serviceText: String = serviceFieldList[position].first
+        val fieldKey = serviceFieldList[position].second
+
+        // Set label for the text view
+        // Reformat text to "Capital Words" intead of "capital_words"
+        dynamicFieldLayout.dynamic_autocomplete_label.text =
+            fieldKey.split("_").map {
+                if (it == "id") it.toUpperCase()
+                else it.capitalize()
+            }.joinToString(" ")
+
+        // If field is looking for an entity_id,
+        // populate the autocomplete with the list of entities
+        if (fieldKey == "entity_id" && entities.isNotEmpty()) {
+            // Only populate with entities for the domain
+            // or for homeassistant domain, which should be able
+            // to manipulate entities in any domain
+            val domain = services[serviceText]!!.domain
+            val domainEntities: ArrayList<String> = ArrayList()
+            if (domain == ("homeassistant")) {
+                domainEntities.addAll(entities.keys)
+            } else {
+                entities.keys.forEach {
+                    if (it.startsWith(domain) || it.startsWith("group")) {
+                        domainEntities.add(it)
+                    }
+                }
+            }
+
+            val adapter = SingleItemArrayAdapter<String>(context) { it!! }
+            adapter.addAll(domainEntities.sorted().toMutableList())
+            autoCompleteTextView.setAdapter(adapter)
+            autoCompleteTextView.onFocusChangeListener = dropDownOnFocus
+        } else if (services[serviceText]!!.serviceData.fields[fieldKey]!!.values != null) {
+            // If a non-"entity_id" field has specific values,
+            // populate the autocomplete with valid values
+            val fieldAdapter = SingleItemArrayAdapter<String>(context) { it!! }
+            fieldAdapter.addAll(
+                services[serviceText]!!.serviceData.fields[fieldKey]!!.values!!.sorted().toMutableList()
+            )
+            autoCompleteTextView.setAdapter(fieldAdapter)
+            autoCompleteTextView.onFocusChangeListener = dropDownOnFocus
+        }
+
+        // Bind the textview to an easily-accessible list for faster fetching of text
+        fieldTextViewList.add(position, autoCompleteTextView)
```

Comment:
When the view actually gets recycled we are going to lose state.  We should add a textChangeListener and store the value on change and when binding the view we should populate with the existing value.

Reply:
The TextWatcher has been implemented in commit 29fc233, along with the changes to how the data is stored.